src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,481 @@
+
+        /// Read map key.
+        String key;
+        readQuotedString(key, buf);
+        assertChar(':', buf);
+        skipWhitespaceIfAny(buf);
+        /// Read map value.
+        String value;
+        readQuotedField(value, buf);
+        assertChar(',', buf);
+        skipWhitespaceIfAny(buf);
+
+        if (key == "descr")
+            descr = value;
+        else if (key == "fortran_order")
+        {
+            if (value != "false")
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Fortran order is not supported");
+        }
+        else if (key == "shape")
+            shape = value;
+    }
+
+    if (shape.empty() || descr.empty())
+        throw Exception(ErrorCodes::INCORRECT_DATA, "npy file header doesn't contain required field 'shape' or 'descr'");
+
+    size_t read_bytes = buf.count() - header_start;
+    if (read_bytes > header_length)
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Header size is incorrect");
+
+    /// Ignore remaining header data.
+    buf.ignore(header_length - read_bytes);
+
+    if (descr[0] == '\'')
+        descr = descr.substr(1, descr.length() - 1);
+    if (descr[descr.length() - 1] == '\'')
+        descr = descr.substr(0, descr.length() - 1);
+
+    if (shape[0] == '\'')
+        shape = shape.substr(1, shape.length() - 1);
+    if (shape[shape.length() - 1] == '\'')
+        shape = shape.substr(0, shape.length() - 1);
+
+    NumpyHeader res;
+    res.shape = parseShape(shape);
+    res.numpy_type = parseType(descr);
+
+    return res;
+}
+
+DataTypePtr getNestedType(DataTypePtr type)
+{
+    String a = type->getName();
+    while (const auto * temp_type = typeid_cast<const DataTypeArray *>(type.get()))
+    {
+        type = temp_type->getNestedType();
+        a = type->getName();
+    }
+
+    return type;
+}
+}
+
+void NpyRowInputFormat::readPrefix()
+{
+    header = parseHeader(*in);
+}
+
+NpyRowInputFormat::NpyRowInputFormat(ReadBuffer & in_, Block header_, Params params_)
+    : IRowInputFormat(std::move(header_), in_, std::move(params_))
+{
+    auto types = getPort().getHeader().getDataTypes();
+    if (types.size() != 1)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected number of columns for Npy input format, expected one column, got {} columns", types.size());
+    nested_type = getNestedType(types[0]);
+}
+
+void NpyRowInputFormat::readRows(MutableColumns & columns)
+{
+    auto & column = columns[0];
+    IColumn * current_column = column.get();
+    size_t elements_in_current_column = 1;
+    for (size_t i = 1; i != header.shape.size(); ++i)
+    {
+        auto & array_column = assert_cast<ColumnArray &>(*current_column);
+        /// Fill offsets of array columns.
+        for (size_t j = 0; j != elements_in_current_column; ++j)
+            array_column.getOffsets().push_back(array_column.getOffsets().back() + header.shape[i]);
+        current_column = &array_column.getData();
+        elements_in_current_column *= header.shape[i];
```

Comment:
User can specify type with wrong Array nesting and in this case we will get an exception from `assert_cast`, but exception message will be not user-friendly. Let's check if column is Array and throw exception with good message. Smth like:
```suggestion
        auto * array_column = typeid_cast<ColumnArray *>(current_column.get());
        if (!array_column)
            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected nesting level of column '{}', expected {}, got {}", column->getName(), header.shape.size() - 1, i - 1);

        /// Fill offsets of array columns.
        for (size_t j = 0; j != elements_in_current_column; ++j)
            array_column.getOffsets().push_back(array_column.getOffsets().back() + header.shape[i]);
        current_column = &array_column.getData();
        elements_in_current_column *= header.shape[i];
```